lib/Dredd.js

```diff
@@ -1,35 +1,94 @@
+const readLocation = require('./readLocation');
+const resolveModule = require('./resolveModule');
 const logger = require('./logger');
-const Runner = require('./TransactionRunner');
+const TransactionRunner = require('./TransactionRunner');
 const { applyConfiguration } = require('./configuration');
-const options = require('./options.json');
 
 
-const PROXY_ENV_VARIABLES = ['HTTP_PROXY', 'HTTPS_PROXY', 'NO_PROXY'];
-const FILE_DOWNLOAD_TIMEOUT = 5000;
+function prefixErrors(decoratedCallback, prefix) {
+  return (error, ...args) => {
+    if (error) { error.message = `${prefix}: ${error.message}`; }
+    decoratedCallback(error, ...args);
+  };
+}
+
+
+function readLocations(locations, options, callback) {
+  if (typeof options === 'function') { [options, callback] = [{}, options]; }
+
+  async.map(locations, (location, next) => {
+    const decoratedNext = prefixErrors(next, `Unable to load API description document from '${location}'`);
+    readLocation(location, options, decoratedNext);
+  }, (error, contents) => {
+    if (error) { callback(error); return; }
+
+    const apiDescriptions = locations
+      .map((location, i) => ({ location, content: contents[i] }));
+    callback(null, apiDescriptions);
+  });
+}
+
+
+function parseContent(apiDescriptions, callback) {
+  async.map(apiDescriptions, ({ location, content }, next) => {
+    const decoratedNext = prefixErrors(next, `Unable to parse API description document '${location}'`);
+    parse(content, decoratedNext);
+  }, (error, parseResults) => {
+    if (error) { callback(error); return; }
+
+    const parsedAPIdescriptions = apiDescriptions
+      .map((apiDescription, i) => Object.assign({}, parseResults[i], apiDescription));
+    callback(null, parsedAPIdescriptions);
+  });
+}
+
+
+function compileTransactions(apiDescriptions) {
+  return apiDescriptions
+    .map(({ mediaType, apiElements, location }) => {
+      try {
+        return compile(mediaType, apiElements, location);
+      } catch (error) {
+        error.message = (
+          'Unable to compile HTTP transactions from '
+          + `API description document '${location}': ${error.message}`
+        );
+        throw error;
+      }
+    })
+    .map((compileResult, i) => Object.assign({}, compileResult, apiDescriptions[i]));
+}
+
+
+function toTransactions(apiDescriptions) {
+  return apiDescriptions
+    // produce an array of transactions for each API description,
+    // where each transaction object gets an extra 'apiDescription'
+    // property with details about the API description it comes from
+    .map(apiDescription => (
+      apiDescription.transactions
+        .map(transaction => Object.assign({
+          apiDescription: {
+            location: apiDescription.location,
+            mediaType: apiDescription.mediaType,
+          },
+        }, transaction))
+    ))
+    // flatten array of arrays
+    .reduce((flatArray, array) => flatArray.concat(array), []);
+}
 
 
 class Dredd {
```

Comment:
What I did here:

- removed stuff, which didn't seem to be necessary anywhere: `this.tests`, `this.stats.fileBasedReporters`
- moved some things to configuration, mostly various coercing and preparation of `this.configuration`, and `this.logProxySettings()`
- moved `start` and `end` events to the transaction runner, as all other events are handled there so I thought it's a good idea to streamline the responsibilities
- I tried to make the constructor free of side-effects and any heavy lifting, and to move all that into `this.run()`

The Dredd class is now responsible for:

- coercing the configuration in the constructor
- requiring a module provided by `--require`
- configure reporters
- "API description locations to transactions" pipeline
- print annotations (and halt Dredd in case of errors)
- launch the transaction runner (where all the testing happens) with the list of transactions